Add one-dimensional Fermi-Hubbard model

bartandrews opened the pull request:
Fixes #102

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/qiskit-community/ffsim?pullRequest=106) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/qiskit-community/ffsim?pullRequest=106) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/qiskit-community/ffsim?pullRequest=106) it.</sub>

bartandrews commented:
I have fixed the lint and format errors (as per the readme instructions), and the tox runs successfully on my computer now

kevinsung commented on .gitignore:
```diff
@@ -105,6 +105,7 @@ celerybeat.pid
 # Environments
 .env
 .venv
+.ffsim
```
Please revert this change, since it seems to be special to your setup. You can add your own git ignore rules to `~/.config/git/ignore` (at least on Linux).

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,79 @@
+    U: float = 0,
+    mu: float = 0,
+    V: float = 0,
+    PBC: bool = False,
```
Instead of single-letter name or acronyms, give these variables descriptive names.

kevinsung replied:
The new names still aren't descriptive enough. Suggestions:
- Instead of `t_hop`, `tunneling`
- Instead of `U_int`, `interaction`
- Instead of `mu_pot`, `chemical_potential`
- Instead of `V_pot`, `nearest_neighbor_interaction`
- Instead of `PBC`, `periodic`

You can also look at the corresponding function from [OpenFermion](https://github.com/quantumlib/OpenFermion/blob/e2fe1e678d973303fab231ce9b3cd83757413b94/src/openfermion/hamiltonians/hubbard.py#L20) for inspiration.

Also, let's not set default values for `t` and `U`.

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,79 @@
+    """
+    coeffs: dict[tuple[tuple[bool, bool, int], ...], complex] = {}
+
+    for p in range(norb - int(not PBC)):
```
I find `range(norb - 1 + PBC)` to be more readable.

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,79 @@
+        coeffs[(cre_b((p + 1) % norb), des_b(p))] = -t
+        coeffs[(cre_a(p), des_a((p + 1) % norb))] = -t
+        coeffs[(cre_b(p), des_b((p + 1) % norb))] = -t
+        if V != 0:
```
```suggestion
        if V:
```

kevinsung replied:
Same comment below.

kevinsung reviewed (changes requested):
Thank you! Please add some tests.

bartandrews commented:
Thank you for the code review! I have implemented your suggestions and have added some tests, which benchmark the ground-state energies for an L=4 chain against known results (from pyblock3, TeNPy, etc.). The tox runs successfully on my computer

kevinsung commented on python/ffsim/operators/__init__.py:
```diff
@@ -32,4 +33,5 @@
     "des_a",
     "des_b",
     "number_operator",
+    "fermi_hubbard",
```
Alphabetical order please.

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,79 @@
+    V_int: float = 0,
+    PBC: bool = False,
+) -> FermionOperator:
+    r"""Fermi-Hubbard operator.
```
```suggestion
    r"""Fermi-Hubbard model Hamiltonian.
```

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,79 @@
+) -> FermionOperator:
+    r"""Fermi-Hubbard operator.
+
+    The Fermi-Hubbard operator for :math:`N` spatial orbitals is defined as
```
```suggestion
    The Hamiltonian for the Fermi-Hubbard model with :math:`N` spatial orbitals is given by
```

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,79 @@
+        PBC: The periodic boundary conditions flag.
+
+    Returns:
+        The Fermi-Hubbard operator for a specified number of orbitals.
```
```suggestion
        The Fermi-Hubbard model Hamiltonian.
```

kevinsung commented on tests/operators/fermi_hubbard_test.py:
```diff
@@ -0,0 +1,130 @@
+    ham = ffsim.linear_operator(op, norb=4, nelec=(2, 2))
+    eigs, _ = scipy.sparse.linalg.eigsh(ham, which="SA", k=1)
+    E = -4.472135955000
+    assert np.isclose(eigs[0], E)
```
Instead of
```python
assert np.isclose(a, b)
```
do
```
np.testing.assert_allclose(a, b)
```
This gives more informative error messages.

kevinsung commented on tests/operators/fermi_hubbard_test.py:
```diff
@@ -0,0 +1,130 @@
+    op = fermi_hubbard(norb=4, t_hop=1)
+    ham = ffsim.linear_operator(op, norb=4, nelec=(2, 2))
+    eigs, _ = scipy.sparse.linalg.eigsh(ham, which="SA", k=1)
+    E = -4.472135955000
```
Variable names should be all lowercase and descriptive (i.e. `energy` instead of `e`. In this case, I would avoid defining a variable altogether and just put the number directly in the assert statement.

bartandrews commented:
Thanks again for the suggested improvements -- I have implemented these changes now

bartandrews commented:
Thanks for catching this, I have implemented these changes now. Also, I want to confirm that in the docstring, when I write `` :math: `\mu` ``, is it okay to use to use a single slash here? Sometimes the backslash needs to be escaped/doubled for proper rendering

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,81 @@
+from ffsim.operators.fermion_action import cre_a, cre_b, des_a, des_b
+
+
+def fermi_hubbard(
```
It just occurred to me that this is only the one-dimensional Fermi-Hubbard model. I might need to rethink the API, because we definitely want to support the 2d model at some point. Let's rename this function to `fermi_hubbard_1d` for now.

bartandrews commented:
Thank you for adding a note on how to compile the docs locally. The mu was indeed rendered correctly, but viewing the docs allowed me to fix a different issue, which was useful. I have renamed to `fermi_hubbard_1d` in anticipation of a `fermi_hubbard_2d` in `fermi_hubbard.py` and corresponding functions `test_fermi_hubbard_2d...` in `fermi_hubbard_test.py`. I can implement that in a separate PR if needed

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,81 @@
+        a^\dagger_{p, \sigma} a^\dagger_{p, \sigma'} a_{p, \sigma'} a_{p, \sigma}
+        - \mu \sum_{p=1}^N \sum_{\sigma} a^\dagger_{\sigma, p} a_{\sigma, p}
+        + V \sum_{p=1}^{N-1} \sum_{\sigma, \sigma'}
+        a^\dagger_{p, \sigma} a^\dagger_{p+1, \sigma'} a_{p+1, \sigma'} a_{p, \sigma}
```
Please rewrite the interaction terms by defining the number operator
```
n_{\sigma, p} = a_{\sigma, p}^\dagger a_{\sigma, p}
```
I think that will make it a bit more readable. Once you do that, make sure the order of the operators in the code matches the order in the documentation. Also, please consistently use (spin, index), i.e. `\sigma, p` instead of `p, \sigma`.

Also, I know the Hubbard model is very common, but it would still be good to have a reference for it.

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,81 @@
+    nearest_neighbor_interaction: float = 0,
+    periodic: bool = False,
+) -> FermionOperator:
+    r"""1D Fermi-Hubbard model Hamiltonian.
```
```suggestion
    r"""One-dimensional Fermi-Hubbard model Hamiltonian.
```

kevinsung commented on tests/operators/fermi_hubbard_test.py:
```diff
@@ -0,0 +1,165 @@
+def test_fermi_hubbard_1d_hermiticity():
+    """Test 1D Fermi-Hubbard model Hamiltonian hermiticity."""
+
+    n_orbitals, n_alpha, n_beta = 4, 3, 1
+    dim = comb(n_orbitals, n_alpha) * comb(n_orbitals, n_beta)
```
```suggestion
    norb = 4
    nelec = (3, 1)
    dim = ffsim.dim(norb, nelec)
```
Sorry for the extreme nitpicking but this is how it's done in the rest of the codebase.

kevinsung commented on tests/operators/fermi_hubbard_test.py:
```diff
@@ -0,0 +1,165 @@
+        nearest_neighbor_interaction=0.4,
+    )
+    ham = ffsim.linear_operator(op, norb=n_orbitals, nelec=(n_alpha, n_beta))
+    np.testing.assert_allclose(ham.dot(np.eye(dim)), ham.H.dot(np.eye(dim)))
```
Instead of `a.dot(b)`, do `a @ b`, it's more readable and consistent with the rest of the codebase.

kevinsung reviewed (changes requested):
The eigenvalue tests are great but I'm trying to think of a test that's more "direct". I think this might work:
1. Create a Fermi-Hubbard model
2. Print its string representation using `repr` and manually check that it makes sense
3. In the test, assert that the string representation is equal to the one obtained in step 2.

bartandrews commented:
Thank you for catching these inconsistencies, I have implemented the changes now :rocket: The next pull request should be smoother once I have learnt your code conventions :+1: (I have also changed the test location to be in the python subdirectory.)

kevinsung commented on tests/python/operators/fermi_hubbard_test.py:
```diff
@@ -0,0 +1,251 @@
+        coeffs[(cre_a(p), des_a((p + 1) % 4))] = -1
+        coeffs[(cre_b(p), des_b((p + 1) % 4))] = -1
+        coeffs[(cre_a((p + 1) % 4), des_a(p))] = -1
+        coeffs[(cre_b((p + 1) % 4), des_b(p))] = -1
```
I had something a bit different in my mind for the test, actually. Please add a single test, similar to [this one](https://github.com/qiskit-community/ffsim/blob/561052c6486e7daf7f828a8d4096eff99f26cc33/tests/python/operators/common_operators_test.py#L21) for `number_operator`. The test should compare the dictionary of the operator with one constructed explicitly without any for loop logic. Then, you can revert these other tests back to how they were originally, where the only thing they did was check the eigenvalue. That gives a better logical separation between what the tests are doing. For good measure, you can rename the tests by appending `_eigenvalue` to the test name and mention it in the docstring, e.g.
```python
def test_non_interacting_fermi_hubbard_1d_eigenvalue():
    """Test non-interacting 1D Fermi-Hubbard model eigenvalues."""
```

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,87 @@
+# (C) Copyright IBM 2023.
```
```suggestion
# (C) Copyright IBM 2024.
```

kevinsung commented on tests/python/operators/fermi_hubbard_test.py:
```diff
@@ -0,0 +1,251 @@
+# (C) Copyright IBM 2023.
```
```suggestion
# (C) Copyright IBM 2024.
```

bartandrews commented:
As requested, in the tests I have now "compared the dictionary of the operator with one constructed explicitly without any for loop logic", for open and periodic boundary conditions. In the process, I also noticed that some existing tests in ffsim use `assert A == B` rather than `np.testing.assert_equal(A, B)`, which seems to go against your convention. Should these also be fixed?

kevinsung commented:
>assert A == B rather than np.testing.assert_equal(A, B)

The reason to use `np.testing.assert_allclose` is that the error message displays the values of the arrays being compared. When A and B aren't arrays then the error messages from pytest are good enough. In any case, unrelated improvements should be done in a separate PR.

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,87 @@
+    Returns:
+        The one-dimensional Fermi-Hubbard model Hamiltonian.
+
+    .. _The Hubbard Model: https://doi.org/10.1146/annurev-conmatphys-031620-102024
```
Thanks for adding this reference. Do you happen to know of one that also includes the nearest-neighbor interaction? I don't think this one does, but perhaps I missed it. I actually haven't seen this term as part of the Hubbard model before.

bartandrews replied:
For the Fermi-Hubbard model, I was following a similar functionality to TeNPy: https://tenpy.readthedocs.io/en/latest/reference/tenpy.models.hubbard.FermiHubbardModel.html#tenpy.models.hubbard.FermiHubbardModel However, the additional V term is fairly innocuous. For example, Eq.1 in the reference doesn't have a chemical potential term either, but this can simply be added, in case it's needed.

bartandrews replied:
Sometimes this is referred to as the "extended Fermi-Hubbard model", e.g. https://arxiv.org/abs/2209.11777

bartandrews replied:
However, it is difficult to find a single reference for this, because you can add on whatever you like to the Fermi-Hubbard model. The V nearest-neighbor term just seemed like it might be useful.

kevinsung replied:
I see, thanks. I think it's fine then.

kevinsung commented on python/ffsim/operators/fermi_hubbard.py:
```diff
@@ -0,0 +1,87 @@
+def fermi_hubbard_1d(
+    norb: int,
+    tunneling: float,
+    interaction: float,
+    chemical_potential: float = 0,
```
```suggestion
    interaction: float,
    *,
    chemical_potential: float = 0,
```
Let's force keyword arguments here so we can add more later without breaking the API.

bartandrews replied:
That's a good idea! I have added this now

kevinsung reviewed (approved):
Thanks Bart!
